chemprop/v2/data/utils.py

```diff
@@ -1,66 +1,163 @@
+        mol_split_fun = train_test_split_molecules
     else:
-        raise ValueError(f"Uknown split type! got: {split}")
+        astartes_kwargs["val_size"] = sizes[1]
+
+    train, val, test = None, None, None
+    match split:
+        case SplitType.CV_NO_VAL, SplitType.CV:
+            if (max_folds := len(data)) > num_folds or num_folds <= 1:
+                raise ValueError(f"Number of folds for cross-validation must be between 2 and {max_folds} (length of data) inclusive (got {num_folds}).")
+
+            train, val, test = [], [], []
+            for _ in range(len(num_folds)):
+                result = split_fun(np.arange(len(data)), sampler="random", **astartes_kwargs)
+                i_train, i_val, i_test = _unpack_astartes_result(data, result, include_val)
+                train.append(i_train)
+                val.append(i_val)
+                test.append(i_test)
+
+        case SplitType.SCAFFOLD_BALANCED:
+            mols_without_atommaps = []
+            for mol_datapoint in data:
+                mol = mol_datapoint.mol
+                copied_mol = copy.deepcopy(mol)
+                for atom in copied_mol.GetAtoms():
+                    atom.SetAtomMapNum(0)
+                mols_without_atommaps.append([copied_mol])
+            result = mol_split_fun(np.array(mols_without_atommaps), sampler="scaffold", **astartes_kwargs)
+            train, val, test = _unpack_astartes_result(data, result, include_val)
+
+        # Use to constrain data with the same smiles go in the same split.
+        case SplitType.RANDOM_WITH_REPEATED_SMILES:
+            # get two arrays: one of all the smiles strings, one of just the unique
+            all_smiles = np.array([d.mol for d in data])
+            unique_smiles = np.unique(all_smiles)
+
+            # save a mapping of smiles -> all the indices that it appeared at
+            smiles_indices = {}
+            for smiles in unique_smiles:
+                smiles_indices[smiles] = np.where(all_smiles == smiles)[0]
+
+            # randomly split the unique smiles
+            result = split_fun(np.arange(len(unique_smiles)))
+            train_idxs, val_idxs, test_idxs = _unpack_astartes_result(None, result, include_val)
 
+            # convert these to the 'actual' indices from the original list using the dict we made
+            train = list(itertools.chain.from_iterable(smiles_indices[unique_smiles[i]] for i in train_idxs))
+            val = list(itertools.chain.from_iterable(smiles_indices[unique_smiles[i]] for i in val_idxs))
+            test = list(itertools.chain.from_iterable(smiles_indices[unique_smiles[i]] for i in test_idxs))
+
+        case SplitType.RANDOM:
+            result = split_fun(np.arange(len(data)), sampler="random", **astartes_kwargs)
+            train, val, test = _unpack_astartes_result(data, result, include_val)
+
+        case SplitType.KENNARD_STONE:
+            result = mol_split_fun(
+                np.array([d.mol for d in data]),
```

Comment:
`astartes` operates on numpy arrays (since the splitting backends actually operate on the values, rather than just the random number generator) and raises warnings about conversions for other data types.